# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/quota/test/xpcshell/test_initializePersistentStorage.js
# Commit: 46a5babaafd2
# Full Hash: 46a5babaafd2c0f2198f353d62254cf518dab391
# Author: Jan Varga <jvarga@mozilla.com>
# Date: 2024-10-15 04:16:11
# Regressor Bug: 1866402
# File Overlap Count: 1
# Description:
#   Bug 1866402 - Implement initialization of entire persistent storage; r=dom-storage-reviewers,jari
#   
#   IndexedDB still calls QuotaManager::EnsurePersistentOriginIsInitializedInternal
#   from its idle database maintenance. However, we want to get rid of that so the
#   method can be made private. The idle database maintenance should only rely on
# ==============================================================================

diff -r 2ff5f3aaf841 -r 46a5babaafd2 dom/quota/test/xpcshell/test_initializePersistentStorage.js
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/dom/quota/test/xpcshell/test_initializePersistentStorage.js	Mon Oct 14 10:06:14 2024 +0000
@@ -0,0 +1,49 @@
+/**
+ * Any copyright is dedicated to the Public Domain.
+ * http://creativecommons.org/publicdomain/zero/1.0/
+ */
+
+/**
+ * This test is mainly to verify initializePersistentStorage() does call
+ * QuotaManager::EnsurePersistentStorageIsInitializedInternal() which does
+ * various things, for example, it restores the directory metadata if it's
+ * broken or missing.
+ */
+
+async function testSteps() {
+  const originDirPath = "storage/permanent/https+++foo.example.com";
+  const metadataFileName = ".metadata-v2";
+
+  info("Initializing");
+
+  let request = init();
+  await requestFinished(request);
+
+  info("Verifying initialization status");
+
+  await verifyInitializationStatus(true, false, false);
+
+  info("Creating an empty directory");
+
+  let originDir = getRelativeFile(originDirPath);
+  originDir.create(Ci.nsIFile.DIRECTORY_TYPE, parseInt("0755", 8));
+
+  info("Initializing persistent storage");
+
+  request = initializePersistentStorage();
+  await requestFinished(request);
+
+  info(
+    "Verifying directory metadata was restored after calling " +
+      "initializePersistentStorage()"
+  );
+
+  let metadataFile = originDir.clone();
+  metadataFile.append(metadataFileName);
+
+  ok(metadataFile.exists(), "Directory metadata file does exist");
+
+  info("Verifying initialization status");
+
+  await verifyInitializationStatus(true, true, false);
+}